zkevm-circuits/src/pi_circuit2.rs

```diff
@@ -380,1147 +220,479 @@ impl<F: Field> SubCircuitConfig<F> for PiCircuitConfig<F> {
-            || Value::known(tx_id),
-        )?;
-        region.assign_advice(
-            || "tx_id_inv",
-            self.tx_id_inv,
-            calldata_offset,
-            || Value::known(tx_id_inv),
-        )?;
-        region.assign_fixed(
-            || "tag",
-            self.tx_table.tag,
-            calldata_offset,
-            || Value::known(tag),
-        )?;
-        region.assign_advice(
-            || "index",
-            self.tx_table.index,
-            calldata_offset,
-            || Value::known(index),
-        )?;
-        region.assign_advice(
-            || "tx_value",
-            self.tx_table.value,
-            calldata_offset,
-            || Value::known(tx_value),
-        )?;
-        region.assign_advice(
-            || "tx_value_inv",
-            self.tx_value_inv,
-            calldata_offset,
-            || Value::known(tx_value_inv),
-        )?;
-        region.assign_advice(
-            || "tx_id_diff_inv",
-            self.tx_id_diff_inv,
-            calldata_offset,
-            || Value::known(tx_id_diff_inv),
-        )?;
-        region.assign_advice(
-            || "is_final",
-            self.is_final,
-            calldata_offset,
-            || Value::known(is_final),
-        )?;
-        region.assign_advice(
-            || "gas_cost",
-            self.calldata_gas_cost,
-            calldata_offset,
-            || Value::known(gas_cost),
-        )?;
-
-        let value_offset = 3 * tx_table_len;
-
-        region.assign_advice(
-            || "raw_pi.tx_value",
-            self.rpi,
-            offset + value_offset + self.circuit_block_len(),
-            || Value::known(tx_value),
-        )?;
+        let randomness = if keccak_hi_lo {
+            challenges.evm_word()
+        } else {
+            challenges.keccak_input()
+        };
+        let rpi = field_bytes
+            .iter()
+            .fold(Value::known(F::zero()), |acc, byte| {
+                acc.zip(t)
+                    .and_then(|(acc, t)| Value::known(acc * t + F::from(*byte as u64)))
+            });
+        let mut cells = vec![None; field_bytes.len() + 2];
+        for (i, byte) in field_bytes.iter().enumerate() {
+            let row_offset = *offset + i;
+
+            region.assign_fixed(
+                || "is_field_rlc",
+                self.is_field_rlc,
+                row_offset,
+                || Value::known(use_rlc),
+            )?;
 
-        // Add copy to vec
-        rpi_vals[offset + value_offset] = Value::known(tx_value);
+            // assign field bytes
+            let field_byte_cell = region.assign_advice(
+                || "field bytes",
+                self.rpi_field_bytes,
+                row_offset,
+                || Value::known(F::from(*byte as u64)),
+            )?;
+            field_rlc_acc = field_rlc_acc
+                .zip(t)
+                .and_then(|(acc, t)| Value::known(acc * t + F::from(*byte as u64)));
```

Comment:
Why the `zip` and `and_then` and not just do the calculation for the single value?

Reply:
`zip` is used for combine two `Value<X>` and `Value<Y>` into `Value<(X, Y)>`
and `Value` has `None` value, so using the `and_then` for calculate the `Value<(X, Y)>` when `Value` is `Some`

Reply:
Oh, I find that the `Value` has implemented the `Add` trait, so just using `+` is OK